python/test_infra/ttir_builder.py

```diff
@@ -4,25 +4,87 @@
 
 from typing import List, Optional, Union, Tuple
 from ttmlir.ir import *
-from ttmlir.dialects import ttir, tt, func
+from ttmlir.dialects import ttir, tt, func, tensor
 from .ttir.ttir_op_test_base import TTIROpTestBase
 from .tensor_builder import TensorBuilder
+from test_infra.ttir import Add
+import torch
 
 
 class TTIRBuilder:
     def __init__(self, ctx: Context, location: Location):
         self.ctx = ctx
         self.loc = location
+
         tt.register_dialect(self.ctx)
         ttir.register_dialect(self.ctx)
 
         self.tensor_builder = TensorBuilder(self.ctx, self.loc)
 
+    def __wrap_operand_constraints(
+        self,
+        num_operands: int,
+        operand_constraints: Optional[List[tt.OperandConstraint]] = None,
+    ) -> tt.ir.OperandConstraintAttr:
+        """
+        Helper method to prepack operand constraints given as a list of enums to a list of
+        tt.ir.OperandConstraintAttr and wraps that list in an tt.ir.OperandConstraintAttr.
+        """
+        operand_constraints = (
+            operand_constraints
+            if operand_constraints is not None
+            else [tt.OperandConstraint.Any for _ in range(num_operands)]
+        )
+
+        return tt.ir.OperandConstraintAttr.get(
+            self.ctx,
+            [
+                tt.ir.OperandConstraintAttr.get(self.ctx, operand_constraint)
+                for operand_constraint in operand_constraints
+            ],
+        )
+
+    @property
+    def __default_dtype(self) -> Type:
+        return F32Type.get(self.ctx)
+
+    def ranked_tensor(
+        self,
+        shape: Union[List[int], Tuple[int]],
+        data_type: Optional[Type] = None,
+        encoding: Optional[Attribute] = None,
+    ) -> RankedTensorType:
+        dtype = data_type if data_type is not None else self.__default_dtype
+        with self.ctx, self.loc:
+            return RankedTensorType.get(shape, dtype, encoding)
+
+    def empty(
+        self,
+        shape: Union[List[int], Tuple[int]],
+        data_type: Optional[Type] = None,
+    ) -> Value:
+        dtype = data_type if data_type is not None else self.__default_dtype
+        with self.ctx, self.loc:
+            return tensor.EmptyOp(shape, dtype)
+
+    def add(self, in0: OpView, in1: OpView) -> OpView:
+        with self.ctx, self.loc:
+            out = self.empty(self.__get_shape(in0))
+
+            op = ttir.AddOp(
+                [self.__get_type(out)],
+                [in0.operation, in1.operation],
+                [out],
+                self.__wrap_operand_constraints(3),
+            )
+
+            return func.ReturnOp(self.__coerce_results(op))
```

Comment:
Is this just a placeholder?  I think we should be returning `op`.

Maybe the test could be written:
```python
def test_add2(builder: TTIRBuilder):
    in0 = builder.empty([32, 32])
    in1 = builder.empty([32, 32])
    add = builder.add(in0, in1)
    print(add)
    builder.finish(add)
```

And then the harness uses the `finish` function to infer what the graph return output is.

Reply:
I still have not completed changes, got some problems along the way, I'll ping you back for another review.

I got a separate branch which leans on this one, going further down to flatbuffer. Will send that on MR after I finish this one.

Reply:
Based on https://github.com/tenstorrent/tt-mlir/pull/854#discussion_r1793330301 , I changed the design drastically. Lets start a new comment thread for it.